File: Python/OneAway.py
#There are three types of edits that can be performed on strings: insert a character,
#remove a character, or replace a character. 
#Given two strings, write a function to check if they are one edit (or zero edits) away
class OneAway:
    def IsOneAway(self, cadena1, cadena2):
        if len(cadena1) == len(cadena2):
            return self.reemplazo(cadena1, cadena2)
        if len(cadena1) + 1 == len(cadena2):
            return self.insertar(cadena1, cadena2)
        if len(cadena2) + 1 == len(cadena1):
            return self.insertar(cadena1, cadena2)
        return False
    
    def insertar(self, cadena1, cadena2):    
        index1 = index2 = 0
        while index1 < len(cadena1):
            if cadena1[index1] != cadena2[index2]:
                if index1 != index2:
                    return False
                index2 += 1
            else:
                index2 += 1
                index1 += 1    
        return True    

    def reemplazo(self, cadena1, cadena2):
        DiffOne = False
        for i in range(len(cadena1)):
            if cadena1[i] != cadena2[i]:
                if DiffOne:
                    return False
                DiffOne = True
        return DiffOne

if __name__ == "__main__":
    oa = OneAway()
    print("Ingrese dos palabras ")
    c1 = input()
    c2 = input()
    print(oa.IsOneAway(c1, c2))

File: Java/OneAway.java
import java.util.Scanner;

/* There are three types of edits that can be performed on strings: insert a character,
remove a character, or replace a character. 
Given two strings, write a function to check if they are one edit (or zero edits) away */

public class OneAway {
    public boolean IsOneAway(String cadena1, String cadena2){
        if(cadena1.length() == cadena2.length())
            return reemplazo(cadena1, cadena2);
        
        if(cadena1.length()+1 == cadena2.length())
            return insertar(cadena1, cadena2);  
            
        if(cadena2.length()+1 == cadena1.length())
            return insertar(cadena2, cadena1); 
        return false;    
    }

    boolean insertar(String cadena1, String cadena2){
        int index1 = 0;
        int index2 = 0;
        while (index1 < cadena1.length()) {
            if (cadena1.charAt(index1) != cadena2.charAt(index2)) {
                if (index1 != index2) {
                    return false;
                }
                index2++;
            } else {
                index1++;
                index2++;
            }
        }
        return true;
    }

    boolean reemplazo(String cadena1, String cadena2){
        boolean findDiff = false;
        for(int i = 0; i < cadena1.length(); i++){
            if(cadena1.charAt(i) != cadena2.charAt(i)){
                if(findDiff){
                    return false;
                }
                findDiff = true;
            }
        }
        return findDiff;
    }

    public static void main(String[] args) {
        Scanner sn = new Scanner(System.in);
        OneAway oa = new OneAway();
        System.out.print("Ingrese dos palabras: ");
        String c1 = sn.nextLine();
        String c2 = sn.nextLine();
        System.out.println(oa.IsOneAway(c1, c2));
    }
}